service/common/src/main/java/org/apache/polaris/service/auth/DefaultActiveRolesProvider.java

```diff
@@ -0,0 +1,105 @@
+ * available roles are active for this request.
+ */
+public class DefaultActiveRolesProvider implements ActiveRolesProvider {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultActiveRolesProvider.class);
+  @Inject Provider<RealmContext> realmContextProvider;
+  @Inject MetaStoreManagerFactory metaStoreManagerFactory;
+  @Inject Provider<PolarisGrantManager> polarisGrantManagerProvider;
+
+  @Override
+  public Set<String> getActiveRoles(AuthenticatedPolarisPrincipal principal) {
+    List<PrincipalRoleEntity> activeRoles =
+        loadActivePrincipalRoles(
+            principal.getActivatedPrincipalRoleNames(),
+            principal.getPrincipalEntity(),
+            metaStoreManagerFactory.getOrCreateMetaStoreManager(realmContextProvider.get()));
+    return activeRoles.stream().map(PrincipalRoleEntity::getName).collect(Collectors.toSet());
+  }
+
+  protected List<PrincipalRoleEntity> loadActivePrincipalRoles(
+      Set<String> tokenRoles, PolarisEntity principal, PolarisMetaStoreManager metaStoreManager) {
+    PolarisCallContext polarisContext = CallContext.getCurrentContext().getPolarisCallContext();
+    PolarisGrantManager.LoadGrantsResult principalGrantResults =
+        polarisGrantManagerProvider.get().loadGrantsToGrantee(polarisContext, principal);
+    polarisContext
+        .getDiagServices()
+        .check(
+            principalGrantResults.isSuccess(),
+            "Failed to resolve principal roles for principal name={} id={}",
+            principal.getName(),
+            principal.getId());
+    if (!principalGrantResults.isSuccess()) {
+      LOGGER.warn(
+          "Failed to resolve principal roles for principal name={} id={}",
+          principal.getName(),
+          principal.getId());
+      throw new NotAuthorizedException("Unable to authenticate");
+    }
+    boolean allRoles = tokenRoles.contains(BasePolarisAuthenticator.PRINCIPAL_ROLE_ALL);
+    Predicate<PrincipalRoleEntity> includeRoleFilter =
+        allRoles ? r -> true : r -> tokenRoles.contains(r.getName());
+    List<PrincipalRoleEntity> activeRoles =
+        principalGrantResults.getGrantRecords().stream()
+            .map(
+                gr ->
+                    metaStoreManager.loadEntity(
+                        polarisContext, gr.getSecurableCatalogId(), gr.getSecurableId()))
+            .filter(PolarisMetaStoreManager.EntityResult::isSuccess)
+            .map(PolarisMetaStoreManager.EntityResult::getEntity)
+            .map(PrincipalRoleEntity::of)
+            .filter(includeRoleFilter)
+            .toList();
+    if (activeRoles.size() != principalGrantResults.getGrantRecords().size()) {
+      LOGGER
+          .atWarn()
+          .addKeyValue("principal", principal.getName())
+          .addKeyValue("scopes", tokenRoles)
+          .addKeyValue("roles", activeRoles)
+          .log("Some principal roles were not found in the principal's grants");
```

Comment:
Is this message helpful for trouble-shooting? Why not log individual failures?

Reply:
I don't expect this to happen often, but including the full set of scopes and roles makes it easy to see where the gap is.